ecosystem/op-deployer.md

```diff
@@ -0,0 +1,163 @@
+deploying and upgrading OP Chains remains a complicated and frustrating process. Users are directed to a tutorial
+with [18 manual steps](https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup) to follow in order
+to create their own testnet. Within Labs, we use a combination of different tools depending on where the chain is to be
+hosted:
+
+- **opc** deploys new chains to our cloud infrastructure.
+- The **compose devnet** uses a bunch of custom Python tooling to deploy a chain locally using `docker-compose`.
+- The **Kurtosis devnet** uses a combination of custom deployer scripts and the fault proof's `getting-started.sh` to
+  deploy a chain to a new enclave.
+- The Interop team now has [their own](https://github.com/ethereum-optimism/optimism/pull/11590) devnet setup to deploy
+  a set of interoperable chains locally.
+
+Notably, with the exception of the Interop devnet none of the solutions above enable deploying multiple L2s to a single
+L1 as part of a single logical Superchain. This is a huge gap in functionality, that will only become more magnified as
+time goes on.
+
+To resolve this problem we need set of well-integrated, modular tools that define the canonical way of deploying a
+chain. Users can then leverage these tools to build their own deployment tooling on top.
+
+Much of what's described below is already implemented as part of `opc`. By extracting these concepts out of `opc`
+and into the monorepo, we can make them more accessible to other teams and users.
+
+# Proposed Solution
+
+The high-level architecture is described
+in [Modular Deployments](https://docs.google.com/document/d/13f8UoO9j05PJdvAZDWuVJWD1UAIEAfb8VbphBnPE3cA/edit#heading=h.jxalqw91xhd0),
+so read that document first if you haven't already. This design doc will focus on the implementation details of the
+architecture it describes.
+
+## Deployment Pipeline
+
+The core of the modular deployment tooling is the concept of a _deployment pipeline_. The pipeline consists of stages
+that each perform a single piece of the deployment. The stages each take an input file, and enrich it with the data
+pertaining to that particular stage of the deployment. This allows the output of upstream stages to be cached.
+
+The diagram below outlines what these stages are:
+
+![](./op-deployer/pipeline.png)
+
+## Deployment Intents
+
+We'll create a new tool called `op-deployer` to orchestrate deployments. Users won't need to interact with the pipeline
+directly. Instead, they'll define a _deployment intent_ which will describe the properties of the chains they want to
+deploy. `op-deployer` will then diff the intent against the current state of the deployment, and run the pipeline to
+ensure that the real deployment matches the intent.
+
+An example deployment intent is below. Deployment intents are encoded as TOML to match what the OP Stack Manager
+expects:
+
+```toml
+[intent]
```

Comment:
Within the intent struct, should we aim to have overlap with the superchain registry config structure? Aligning formats between these two tooling could enhance usability and impact by allowing any operator to easily deploy another chain

Reply:
The deployer can output a deploy config, so it's compatible with the registry. Over time I expect us to refactor away from the deploy config, however, so at that time we'll need to refactor the SR as well.